alf/networks/s5/s5.py

```diff
@@ -0,0 +1,421 @@
+        B_bar = ((1 / Lambda) * (Lambda_bar - 1))[:, None] * B_tilde
+        B_bar = _as_real_to_complex_matrix(B_bar)
+        return Lambda_bar, B_bar
+
+    def forward(self, input, state):
+        if input.ndim == 3:
+            return self.forward_sequence(input, state)
+        Lambda_bar, B_bar = self.discretize()
+        cstate = _view_as_complex(state)
+        state = _view_as_real(Lambda_bar * cstate) + input @ B_bar.T
+        output = 2 * state @ self._C.T
+        output = torch.addcmul(output, self._D, input)
+        return output, state
+
+    def forward_sequence(self, inputs, state):
+        r"""
+
+        Args:
+            inputs (torch.Tensor): shape is [T, B, data_dim]
+            state (torch.Tensor): :math:`s_{-1}`. The shape is [batch_size, state_dim]
+        Returns:
+            outputs (torch.Tensor): shape is [T, B, data_dim]
+            state (torch.Tensor): :math:`s_{T-1}`. The shape is [batch_size, state_dim]
+        """
+        Lambda_bar, B_bar = self.discretize()
+        x = _view_as_complex(inputs @ B_bar.T)
+        cstate = _view_as_complex(state)
+        cstates = diag_ssm_forward(cstate, x, Lambda_bar)
+        outputs = 2 * _view_as_real(cstates) @ self._C.T
+        outputs = torch.addcmul(outputs, self._D, inputs)
+        return outputs, _view_as_real(cstates[-1])
+
+
+class BatchNorm1dChannelLast(nn.BatchNorm1d):
+    """Batch normalization layer for channel-last tensor
+    """
+
+    def forward(self, x):
+        """
+        Args:
+            x (torch.Tensor): shape is [T, B, C] or [B, C]
+        """
+        if x.ndim == 2:
+            return super().forward(x)
+        elif x.ndim == 3:
+            x = x.transpose(1, 2)
+            x = super().forward(x)
+            return x.transpose(1, 2)
+        else:
+            raise ValueError("Only support 2d or 3d tensor")
+
+
+class Dropout(nn.Module):
+    """Dropout layer supporting shared dropout mask across given dimension
+
+    Args:
+        dropout (float): dropout rate
+        broadcast_dims (tuple): dimensions to share dropout mask
+    """
+
+    def __init__(self, dropout: float, broadcast_dims: Tuple[int, ...] = ()):
+        super().__init__()
+        self._broadcast_dims = broadcast_dims
+        self._dropout = dropout
+
+    def forward(self, x):
+        if self.training:
+            mask_shape = list(x.shape)
+            for d in self._broadcast_dims:
+                mask_shape[d] = 1
+            mask = torch.rand(mask_shape, device=x.device) > self._dropout
+            return x * mask
+        else:
+            return x * (1 - self._dropout)
+
+
+class S5Block(Network):
+    """
+    Corresponds to s5.layers.SequenceLayer
+
+    Note: for sequence input, the input is assumed to be [T, B, C]
+
+    """
+
+    def __init__(self,
+                 ssm_ctor: Callable[..., nn.Module],
+                 dropout: float,
+                 activation: str = "half_glu2",
```

Comment:
Need to explain the activation here, e.g. what are the possible options and their effects.

Reply:
added.